### PR TITLE
[Dashboard] Sales on dashboard updated

### PR DESCRIPTION
**Pull request by AdamKasp**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

Screenshot 2020-02-07 at 12 06 39

### CODE REVIEW

File: src/Sylius/Behat/Context/Setup/OrderContext.php [hunk 1]
```diff
@@ -942,6 +955,31 @@ private function createOrders(
...
 
+    private function createPaidOrders($numberOfCustomers, $numberOfOrders, $total): void
+    {
+        $customers = $this->generateCustomers($numberOfCustomers);
+        $sampleProductVariant = $this->sharedStorage->get('variant');
+        $total = $this->getPriceFromString($total);
+
+        for ($i = 0; $i < $numberOfOrders; ++$i) {
+            $order = $this->createOrder($customers[random_int(0, $numberOfCustomers - 1)], '#' . uniqid());
+            $order->setState(OrderInterface::STATE_NEW); // Temporary, we should use checkout to place these orders.
```
Comment (Tomanhez) [inline src/Sylius/Behat/Context/Setup/OrderContext.php]:
```suggestion
            $order->setState(OrderInterface::STATE_NEW);
```

File: src/Sylius/Component/Core/Repository/OrderRepositoryInterface.php [hunk 2]
```diff
@@ -52,8 +52,12 @@ public function findLatestCartByChannelAndCustomer(ChannelInterface $channel, Cu
 
     public function getTotalSalesForChannel(ChannelInterface $channel): int;
 
+    public function getTotalPaidSalesForChannel(ChannelInterface $channel): int;
+
     public function countFulfilledByChannel(ChannelInterface $channel): int;
 
+    public function countPaidByChannel(ChannelInterface $channel): int;
+
     /**
```
Comment (Tomanhez) [inline src/Sylius/Component/Core/Repository/OrderRepositoryInterface.php]: Wrong order

File: src/Sylius/Behat/Context/Setup/OrderContext.php (hunk 1 above)
Comment (AdamKasp) [reply src/Sylius/Behat/Context/Setup/OrderContext.php]: but it is true :( and we have this info in another function too

File: src/Sylius/Component/Core/Repository/OrderRepositoryInterface.php (hunk 2 above)
Comment (AdamKasp) [reply src/Sylius/Component/Core/Repository/OrderRepositoryInterface.php]: in interfaces, we don't have any rules for order of functions so I don't know what order will be fine :D

File: src/Sylius/Behat/Context/Setup/OrderContext.php (hunk 1 above)
Comment (GSadee) [reply src/Sylius/Behat/Context/Setup/OrderContext.php]: Could you use state machine here?

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/OrderRepository.php [hunk 3]
```diff
@@ -250,6 +263,19 @@ public function countFulfilledByChannel(ChannelInterface $channel): int
...
     }
 
+    public function countPaidByChannel(ChannelInterface $channel): int
+    {
+        return (int) $this->createQueryBuilder('o')
+            ->select('COUNT(o.id)')
+            ->andWhere('o.channel = :channel')
+            ->andWhere('o.paymentState = :state')
+            ->setParameter('channel', $channel)
+            ->setParameter('state', OrderPaymentStates::STATE_PAID)
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/OrderRepository.php]: I guess we should count `paid` and `fulfilled` orders

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/OrderRepository.php [hunk 4]
```diff
@@ -234,6 +234,19 @@ public function getTotalSalesForChannel(ChannelInterface $channel): int
...
     }
 
+    public function getTotalPaidSalesForChannel(ChannelInterface $channel): int
+    {
+        return (int) $this->createQueryBuilder('o')
+            ->select('SUM(o.total)')
+            ->andWhere('o.channel = :channel')
+            ->andWhere('o.paymentState = :state')
+            ->setParameter('channel', $channel)
+            ->setParameter('state', OrderPaymentStates::STATE_PAID)
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/OrderRepository.php]: I guess we should take into account also `fulfilled` orders

File: src/Sylius/Behat/Context/Setup/OrderContext.php [hunk 5]
```diff
@@ -932,6 +943,7 @@ private function createOrders(
             $this->addVariantWithPriceToOrder($order, $sampleProductVariant, $price);
 
             if ($isFulfilled) {
+                $order->setPaymentState(OrderPaymentStates::STATE_PAID);
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Setup/OrderContext.php]: Never, ever call `set*State`. It should be triggered by state machine.

File: src/Sylius/Behat/Context/Setup/OrderContext.php [hunk 6]
```diff
@@ -932,6 +943,7 @@ private function createOrders(
...
                 $this->applyTransitionOnOrder($order, OrderTransitions::TRANSITION_FULFILL);
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Setup/OrderContext.php]: Should we consider a guard for payment and shipment state on order? /cc @Sylius/core-team

File: src/Sylius/Behat/Context/Setup/OrderContext.php (hunk 1 above)
Comment (lchrusciel) [reply src/Sylius/Behat/Context/Setup/OrderContext.php]: You may try to use OrderExampleFactory here.

File: src/Sylius/Behat/Context/Setup/OrderContext.php [hunk 7]
```diff
@@ -942,6 +954,31 @@ private function createOrders(
...
+            $order = $this->createOrder($customers[random_int(0, $numberOfCustomers - 1)], '#' . uniqid());
+            $order->setState(OrderInterface::STATE_NEW); // Temporary, we should use checkout to place these orders.
+            $this->applyPaymentTransitionOnOrder($order, PaymentTransitions::TRANSITION_COMPLETE);
+
+            $price = $i === ($numberOfOrders - 1) ? $total : random_int(1, $total);
+            $total -= $price;
+
+            $this->addVariantWithPriceToOrder($order, $sampleProductVariant, $price);
+
+           $order->setPaymentState(OrderPaymentStates::STATE_PAID);
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Setup/OrderContext.php]: Never, ever call `set*State`. It should be triggered by state machine.

File: src/Sylius/Behat/Context/Setup/OrderContext.php [hunk 8]
```diff
@@ -964,6 +1001,7 @@ private function createOrdersWithProduct(
             $this->addVariantWithPriceToOrder($order, $sampleProductVariant, $price);
 
             if ($isFulfilled) {
+                $order->setPaymentState(OrderPaymentStates::STATE_PAID);
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Setup/OrderContext.php]: Never, ever call `set*State`. It should be triggered by state machine.

File: src/Sylius/Behat/Context/Setup/OrderContext.php (hunk 1 above)
Comment (Zales0123) [reply src/Sylius/Behat/Context/Setup/OrderContext.php]: There is no reason in Heaven and Earth that should excuse using a `setState` method - and for sure it should be eradicated whenever it's used right now 🔥

File: src/Sylius/Behat/Context/Setup/OrderContext.php (hunk 6 above)
Comment (Zales0123) [reply src/Sylius/Behat/Context/Setup/OrderContext.php]: Agreed, "fulfill" transition should never be called explicitly 🐎

File: src/Sylius/Behat/Context/Setup/OrderContext.php (hunk 7 above)
Comment (Zales0123) [reply src/Sylius/Behat/Context/Setup/OrderContext.php]: + wrong indent 💃

File: src/Sylius/Component/Core/Repository/OrderRepositoryInterface.php (hunk 2 above)
Comment (Zales0123) [reply src/Sylius/Component/Core/Repository/OrderRepositoryInterface.php]: The order of methods should be... rational :D It's seems good IMO

File: src/Sylius/Behat/Context/Setup/OrderContext.php [hunk 9]
```diff
@@ -955,7 +993,7 @@ private function createOrdersWithProduct(
 
         for ($i = 0; $i < $numberOfOrders; ++$i) {
             $order = $this->createOrder($customers[random_int(0, $numberOfCustomers - 1)], '#' . uniqid(), $product->getChannels()->first());
-            $order->setState(OrderInterface::STATE_NEW);
+            $this->stateMachineFactory->get($order, OrderTransitions::GRAPH)->apply(OrderTransitions::TRANSITION_CREATE);
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Setup/OrderContext.php]:
```suggestion
            $this->applyTransitionOnOrder($order, OrderTransitions::TRANSITION_CREATE);
```
?

**Comment by GSadee** [comment]
Thank you, Adam! :1st_place_medal:

**Pull request merged**